### PR TITLE
RABFE

### PR DESCRIPTION
**Pull request by proteneer**
This PR puts everything together and implements the inference version of the RABFE code.

In particular, this PR has several key things of note:

1) The estimated free energy is an "absolute" value that is shifted by an unknown constant (fixed for a pair of "blocker" and "target).
2) The blocker molecule should ideally be the "most stable" pose, and one heuristic we use is to pick the largest molecule in the set if the true reference is unavailable.
3) Both the complex and solvent stages implement an initial parameter conversion stage that allows the states to be "forcefield independent", such that when we retrain, we only have to re-train the easier to converge conversion stage as its fully interacting.
4) The complex stages use a more sophisticated pre-equilibration approach:
- We pre-equilibrate the holo structure with the blocker, but the coordinates will shift over time.
- We then manually rigidly align each ligand to the blocker molecule using the same idxs as that of the core restraints.
- This has has the advantage that we can use the *same* roughly equilibrated starting structure for every molecule for a given target.

### CODE REVIEW

File: examples/validate_relative_binding.py [hunk 1]
```diff
@@ -0,0 +1,347 @@
+# This script runs the rabfe code in "inference" mode.
```
Comment (badisa) [inline examples/validate_relative_binding.py:1]: Is relative binding another term for rabfe?

File: md/minimizer.py [hunk 2]
```diff
@@ -160,3 +176,90 @@ def minimize_host_4d(mols, host_system, host_coords, ff, box) -> np.ndarray:
         assert np.all(norm < 25000)
 
     return final_coords[:num_host_atoms]
+
+
+def equilibrate_complex(
```
Comment (badisa) [inline md/minimizer.py:181]: As a note we should probably look at the equilibrate function I added recently and see if we can't merge this functionality

**Review by badisa — APPROVED** [review]

**Review by maxentile — APPROVED** [review]

File: examples/validate_relative_binding.py (hunk 1 above)
Comment (proteneer) [reply examples/validate_relative_binding.py:1]: This is really an absolute method

File: examples/validate_relative_binding.py [hunk 3]
```diff
@@ -0,0 +1,347 @@
...
+
+import multiprocessing
+from training.dataset import Dataset
+from rdkit import Chem
+
+from timemachine.potentials import rmsd
+from md import builders, minimizer
+
+array = Union[np.array, jnp.array]
+Handler = Union[AM1CCCHandler, LennardJonesHandler] # TODO: do these all inherit from a Handler class already?
```
[the remark this answers is not in the document]
Comment (proteneer) [reply examples/validate_relative_binding.py:45]: willfix

File: examples/validate_relative_binding.py [hunk 4]
```diff
@@ -0,0 +1,347 @@
...
+
+if __name__ == "__main__":
+
+    multiprocessing.set_start_method('spawn')
+
+    parser = argparse.ArgumentParser(
+        description="Absolute Binding Free Energy Testing",
```
[the remark this answers is not in the document]
Comment (proteneer) [reply examples/validate_relative_binding.py:52]: willfix

File: examples/validate_relative_binding.py [hunk 5]
```diff
@@ -0,0 +1,347 @@
...
+        complex_topology,
+        temperature,
+        pressure,
+        dt,
+        cmd_args.num_complex_equil_steps,
+        cmd_args.num_complex_prod_steps
+    )
+
+    # solvent models.
+    solvent_conversion_schedule = np.linspace(0, 1.0, cmd_args.num_solvent_conv_windows)
```
[the remark this answers is not in the document]
Comment (proteneer) [reply examples/validate_relative_binding.py:226]: willfix

File: examples/validate_relative_binding.py [hunk 6]
```diff
@@ -0,0 +1,347 @@
...
+        
+        num_complex_atoms = complex_coords.shape[0]
+
+        # Use core_idxs to generate
+        R, t = rmsd.get_optimal_rotation_and_translation(
+            x1=complex_ref_x0[num_complex_atoms:][core_idxs[:, 1]], # reference core atoms
+            x2=mol_coords[core_idxs[:, 0]], # mol core atoms
+        )
+
+        aligned_mol_coords = rmsd.apply_rotation_and_translation(mol_coords, R, t)
```
[the remark this answers is not in the document]
Comment (proteneer) [reply examples/validate_relative_binding.py:272]: x1 is the blocker ligand, and will be pre-equilibrated (and therefore drift with the host). x2 is the variable ligand in the dataset, so we need to re-align it to blocker at the start of production.

`t` is the translation vector whose tail is at `x1` and head is at `x2`.  `apply_rotation_and_translation` does:

""" python
def apply_rotation_and_translation(x, R, t):
    x_com = np.mean(x, axis=0)
    aligned_x = (x - x_com)@R - t + x_com
    return aligned_x
"""

Which offsets `x2` by `-t` so it's shifted to where `x1` is.

File: md/minimizer.py (hunk 2 above)
Comment (proteneer) [reply md/minimizer.py:181]: the pre-equilibration protocol is slightly different for RABFE due to the blocker alignment, but yes we should still try and merge these two bits of code.

File: timemachine/potentials/rmsd.py [hunk 7]
```diff
@@ -30,6 +30,37 @@ def get_optimal_rotation(x1, x2):
 
     return rotation
 
+def get_optimal_translation(x1, x2):
```
[the remark this answers is not in the document]
Comment (proteneer) [reply timemachine/potentials/rmsd.py:33]: Willfix.
